### PR TITLE
feat(pymongo): add PyMongo integration

### LINKED ISSUE #1585: pymongo integration
### Discussed in https://github.com/getsentry/sentry-python/discussions/1580

<div type='discussions-op-text'>

<sup>Originally posted by **abhishek17569** August 24, 2022</sup>
Hello All,
I'm using pymongo and performing multiple aggregation pipelines. 
I want to get the DB process spans to be shown in event breakdown.

### PR DESCRIPTION
**Pull request by Agalin**
Adds breadcrumbs and performance traces for PyMongo queries using an
official monitoring API. Integration is similar to the one available in
OpenTelemetry, tags set to values recommended for attributes by OT as
specified in `Span Operations` guidelines.

PyMongo version selection explanation:
* 3.1 - introduction of monitoring API. Only Python 2.7 and 3.6
supported.
* 3.12 - latest 3.x release, support for 2.7, 3.6-3.9 (3.7-3.9 added in
various minor releases between 3.1 and 3.12).
* 4.0 - no support for 2.7, added support for 3.10.
* 4.1 - no support for 3.6.0-3.6.1.
* 4.2 - no support for any 3.6.

Uses `MockupDB` for tests (Python implementation of MongoDB protocol server). `mongomock` is a reimplementation of the `MongoClient` and doesn't support tracing. This way we get real world like results. Note that pymongo has changed queries sent for some requests between 3.1 and 3.12 (probably 3.7 or 3.11) so additional responder is necessary to handle tests with 3.1.

There is no support for GridFS nor server connect/disconnect events. Those can be added in the future although I don't think there is an API similar to `pymongo.monitoring` for GridFS (so monkeypatching would be necessary in this one).

closes #1585

### CODE REVIEW

**Comment by Agalin** [comment]
BTW I really hope spans like `db.find` are visible in DB spans panel and similar places, didn't have time to dive into implementation details of span retrieval endpoints.

**Comment by Agalin** [comment]
Docs PR: https://github.com/getsentry/sentry-docs/pull/5469

File: sentry_sdk/integrations/pymongo.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+from sentry_sdk.integrations import DidNotEnable, Integration
+from sentry_sdk.tracing import Span
+from sentry_sdk.utils import capture_internal_exceptions
+
+from sentry_sdk._types import MYPY
+
+try:
+    from pymongo import monitoring
+except ImportError:
+    raise DidNotEnable("Pymongo not installed")
```
Comment (polina-popova) [inline sentry_sdk/integrations/pymongo.py:13]: Hi,
May be would be nice also to explicitly check the pymongo version and give the more specific alert as in SqlalchemyIntegration, for example, because pymongo could be installed but has the wrong version.

File: sentry_sdk/integrations/pymongo.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            data["operation_ids"]["operation"] = event.operation_id
+            data["operation_ids"]["request"] = event.request_id
+
+            try:
+                lsid = command.pop("lsid")["id"]
+                data["operation_ids"]["session"] = str(lsid)
+            except KeyError:
+                pass
+
+            query = "{} {}".format(event.command_name, command)
```
Comment (polina-popova) [inline sentry_sdk/integrations/pymongo.py:73]: I checked this implementation with my project and found

`"find {'find': 'users', 'filter': {'_id': '1'}, 'limit': 1, 'singleBatch': True, '$readPreference': {'mode': 'primary'}}"` and
`'saslStart {}'` as a query values.
So there are double mentioning of the command name and empty curve brackets in case of no `command` value.

Are this results the same as you expected?

Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:73]: It's the expected result. It's command + its arguments. As you can see in tests content of the arguments dict varies between mongo client (protocol) versions and command being repeated here is not always the case.

One place where there is an explicit duplication of command is op name - and I'm still not sure if it should be repeated there or only one of those should stay (and which one).

File: sentry_sdk/integrations/pymongo.py (hunk 1 above)
Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:13]: I can add it although barely any integration does it - even new ones, like FastAPI just check for the import itself.

Comment (polina-popova) [reply sentry_sdk/integrations/pymongo.py:13]: I am not sure that FastAPI has any restrictions in its version to be able to use the integration. Anyway, It was just a proposal.

Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:13]: It does - otherwise there would be no minimal version specified in `setup.py`. For example see https://github.com/getsentry/sentry-python/pull/1349 - FastAPI `0.74.1` had dedicated changes to enable transaction name retrieval. And I believe there were further changes in it to improve integration.

**Comment by antonpirker** [comment]
Hey @Agalin and @polina-popova !
Sorry for the delay! And thank you soo much for the great work! 
I will have a look at this today, and I will also be the person to help bring this into a release sooner than later :-)

Could you maybe rebase your branch again on the newest master, so everything is up to date? Thanks!

**Comment by Agalin** [comment]
Awesome. Rebased to master (clean rebase as the only existing file is tox.ini)

File: sentry_sdk/integrations/pymongo.py (hunk 1 above)
Comment (antonpirker) [reply sentry_sdk/integrations/pymongo.py:13]: Hey! I guess the import check as you implemented is totally fine.

File: sentry_sdk/integrations/pymongo.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            return
+        with capture_internal_exceptions():
+            command = dict(event.command)
+
+            command.pop("$db", None)
+            command.pop("$clusterTime", None)
+            command.pop("$signature", None)
+
+            if event.command_name:
+                op = "db." + event.command_name
```
Comment (antonpirker) [inline sentry_sdk/integrations/pymongo.py:46]: As the span `op` please use "db.query" instead of "db.find". (I know it messes up this elegant code but it helps us figure out performance problems with queries)

File: sentry_sdk/integrations/pymongo.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        # type: (CommandFailedEvent | CommandStartedEvent | CommandSucceededEvent) -> int
+        return event.request_id
+
+    def started(self, event):
+        # type: (CommandStartedEvent) -> None
+        hub = Hub.current
+        if hub.get_integration(PyMongoIntegration) is None:
+            return
+        with capture_internal_exceptions():
+            command = dict(event.command)
```
Comment (antonpirker) [inline sentry_sdk/integrations/pymongo.py:39]: There is a function called `sentry_sdk.hub._should_send_default_pii()` that should be used to guard personal identifiable information. (by default this function returns `False`, this behavior can be changed with this option: https://docs.sentry.io/platforms/python/configuration/options/#send-default-pii)

Right now if you do inserts or make queries, all the user data is sent to Sentry. This must not be the case unless `_should_send_default_pii()` returns `True`.

Can you please make sure that all the values in the command are replaced by `%s`? 
(Maybe there is a way to get the redacted/parameterized command from MongoDB?)

**Review by antonpirker — CHANGES_REQUESTED** [review]
Hey! 

Extraordinary work! 
Just one change to make sure the personal data of users is save and one change to make it easier for Sentry to identify performance problems with queries.

**Comment by antonpirker** [comment]
Oh, and please fix also the linting errors. Thanks!

File: sentry_sdk/integrations/pymongo.py (hunk 3 above)
Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:46]: Nah, it's fine and actually solves a discussion above if `command_name` makes sense being duplicated in `op` and `description`.

Comment (cleptric) [reply sentry_sdk/integrations/pymongo.py:46]: We should try to stay in sync with the list of operations outlined at https://develop.sentry.dev/sdk/performance/span-operations/#database
We just spend a significant amount of time unifying span ops across our SDKs.

File: sentry_sdk/integrations/pymongo.py (hunk 4 above)
Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:39]: There is no such option. 😞 Those aren't events sent to the database but objects crafted specifically to be used by monitoring components and there is no way from there to ensure access the original query object - which doesn't solve the problem either as it's a wrapper around the raw BSON data. There is no SQL-like convention of parameter replacement in mongo.
I don't think there is any clean way to only substitute user-provided data either, especially when you consider protocol changes between pymongo versions.

Basically I can do one of the two things:
* Only add `command` part when PII can be sent.
* Replace values in this dict by `%s` (recursively?).

TBH I'd rather do the first one to avoid performance penalty.

Comment (antonpirker) [reply sentry_sdk/integrations/pymongo.py:39]: I guess doing suggestion 2 would be cool. But I do not think we need recursion right now.
Just something like this should be ok:
```
>>> bla = {
...   "blub": "hello",
...   "foo": {
...     "bar": "hello2",
...   }
... }
>>> bla.keys()
dict_keys(['blub', 'foo'])
>>> for key in bla.keys():
...   bla[key] = "%s"
... 
>>> bla
{'blub': '%s', 'foo': '%s'}
```

If we find out that this is not enough data to debug queries (or just know what queries are run) we can change them in the future.

Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:41]: I've implemented the first one for now.

File: sentry_sdk/integrations/pymongo.py (hunk 3 above)
Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py]: Yeah, I've seen that change. I believe this doc looked a bit different when I was writing the pull request. 😄 Now it's a bit cleaner.

File: sentry_sdk/integrations/pymongo.py (hunk 4 above)
Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:41]: Ah. Missed your comment. Will do it like suggested then.

Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:41]: Added top level replacement.

Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:41]: Oh great. Missed tests update.

BTW should I add tests for that PII part?

Comment (antonpirker) [reply sentry_sdk/integrations/pymongo.py:41]: Tests for PII stuff would be great, yes!

Comment (Agalin) [reply sentry_sdk/integrations/pymongo.py:41]: Added.

**Comment by antonpirker** [comment]
I have now added a new function for nicer PII stripping, could you review my PR @Agalin : https://github.com/operasoftware/sentry-python/pull/1 
Thank you very much!

**Comment by antonpirker** [comment]
I have to stay home to take care of my daughter today (some childrens sickness) Could you fix the tests please @Agalin ? I think when the tests are green we are ready to merge.

**Comment by Agalin** [comment]
@antonpirker tests now pass in tox on my machine (tested 2.7 with pymongo 3.1 and 3.8 with 4.2).

**Comment by antonpirker** [comment]
Amazing @Agalin ! Thank you so much again for your great work!

**Review by antonpirker — APPROVED** [review]

**Comment by antonpirker** [comment]
Hey @Agalin if you now rebase on the new master again, it will be merged automatically (after running the tests once again)

**Comment by Agalin** [comment]
Ah. My rebase stopped automerge. 😓 @antonpirker can you reapprove?

**Comment by antonpirker** [comment]
Done! (now for the documentation)

**Comment by antonpirker** [comment]
Hey everyone!

This will be released today!

Again: thank you so much for your contribution, it is really great work!
@Agalin If you send me your shipping address to anton.pirker@sentry.io I can send you a little "thank you" gift.

**Comment by Agalin** [comment]
Yay, it's happening!
Should I add an entry in the main Sentry codebase that adds integration recommendation? It couldn't be done earlier as version number is required.
